modules/@apostrophecms/ui/ui/apos/stores/modal.js

```diff
@@ -0,0 +1,173 @@
+import { defineStore } from 'pinia';
+import { ref, computed } from 'vue';
+import cuid from 'cuid';
+
+export const useModalStore = defineStore('modal', () => {
+  const stack = ref([]);
+  const activeId = ref(null);
+
+  const activeModal = computed(() => {
+    return stack.value.find(modal => activeId.value === modal.id);
+  });
+
+  function add(modal) {
+    stack.value.push(modal);
+  }
+
+  function get(id) {
+    return stack.value.find(modal => id === modal.id);
+  }
+
+  function getAt(index) {
+    const last = stack.value.length - 1;
+    const target = index < 0
+      ? last + 1 + index
+      : index > stack.value.length
+        ? last
+        : index;
+
+    return stack.value[target] || {};
+  }
+
+  function getProperties(id) {
+    const stackModal = stack.value.find(modal => id === modal.id);
+    if (!stackModal || !apos.modal.modals) {
+      return {};
+    }
+
+    const properties = {
+      ...apos.modal.modals
+        .find(modal => modal.componentName === stackModal.componentName &&
+          modal.props.moduleName === stackModal.props.moduleName)
+    };
+
+    return properties;
+  }
+
+  async function execute(componentName, props) {
+    return new Promise((resolve) => {
+      const item = {
+        id: `modal:${cuid()}`,
+        componentName,
+        resolve,
+        props: props || {},
+        elementsToFocus: [],
+        focusedElement: null
+      };
+
+      activeId.value = item.id;
+      stack.value.push(item);
+      apos.bus.$emit('modal-launched', item);
+    });
+  }
+
+  function updateModalData(id, data) {
+    stack.value = stack.value.map((modal) => {
+      return modal.id === id ? {
+        ...modal,
+        ...data
+      } : modal;
+    });
+  }
+
+  function setModalResult(modalId, result) {
+    stack.value = stack.value.map((modal) => {
+      return modal.id === modalId
+        ? {
+          ...modal,
+          result
+        }
+        : modal;
+    });
+  }
+
+  function resolve(modal) {
+    stack.value = stack.value.filter(_modal => modal.id !== _modal.id);
+    const current = getAt(-1);
+    activeId.value = current.id || null;
+    modal.resolve(modal.result);
+    apos.bus.$emit('modal-resolved', modal);
+  }
+
+  async function confirm(content, options = {}) {
+    return execute(apos.modal.components.confirm, {
+      content,
+      mode: 'confirm',
+      options
+    });
+  }
+
+  async function alert(alertContent, options = {}) {
+    return execute(apos.modal.components.confirm, {
+      content: alertContent,
+      mode: 'alert',
+      options
+    });
+  }
+
+  // Returns true if el1 is "on top of" el2 in the
```

Comment:
I think this function is never used..
modals use the overlay click event to close themselves.

Reply:
Look at `AposTiptapAnchor.vue`:

```
      v-click-outside-element="close"
```

This is used to edit links in rich text. Needs to keep working, so `click-outside-element.js` still needs to work, but could be factored to use the store.